InciCsvCorrector.java

```diff
@@ -0,0 +1,159 @@
+            needsCorrection = true;
+            System.out.print("invalid lines : ");
+            while(!invLines.isEmpty()){
+                System.out.print(invLines.remove(0) + " ");
+            }
+            System.out.println();
+        }
+        if(needsCorrection){
+            try {
+                correctInterruptedLines(source, corrected);
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+            System.out.println("the file has been corrected");
+        }else{
+            System.out.println("the file does not need correction");
+        }
+    }
+
+    /**
+     * finds the invalid lines in the source file and prints them
+     * @param source source file
+     * @return an ArrayList of integers with the line numbers in the source file
+     * @throws IOException when the file does not exist
+     */
+    private static ArrayList<Integer> findInvalidLines(File source) throws IOException {
+
+        BufferedReader br = new BufferedReader(new FileReader(source));
+        String line;
+        int lineNum=0;
+        ArrayList<Integer> invalidLines = new ArrayList<>();
+
+        while (((line = br.readLine()) != null) ) {
+            lineNum++;
+            if(line.length()>MINIMUM_VALID_LENGTH) {
+                boolean currentLineHasValidStart = (isNumeric(line.substring(0, 5)) && (line.charAt(5) == ','));        //valid start of line (12345,)
+                boolean currentLineHasValidEnd = isDate(line.substring(line.length() - DATE_LENGTH));              //valid end of line (dd/mm/yyyy)
+                if (!(currentLineHasValidStart && currentLineHasValidEnd)) {               //adds to list the invalid lines
+                    invalidLines.add(lineNum);
+                }
+            }else{
+                invalidLines.add(lineNum);
+            }
+        }
+        return invalidLines;
+    }
+
+    /**
+     * eliminates line endings that break the ingredient entries writing the results in the corrected file
+     * @param source source file
+     * @param corrected result file
+     */
+    private static void correctInterruptedLines(File source,File corrected) throws IOException {
+
+        BufferedReader br = new BufferedReader(new FileReader(source));
+        BufferedWriter bw = new BufferedWriter(new FileWriter(corrected));
+        String line;
+        boolean currentLineHasValidStart;
+        boolean currentLineHasValidEnd;
+        boolean lineCompleted = true;
+
+        while((line = br.readLine())!=null){
+            if(line.length()>MINIMUM_VALID_LENGTH) {
+                currentLineHasValidStart = (isNumeric(line.substring(0, 5)) && (line.charAt(5) == ','));        //valid start of line (12345,)
+                currentLineHasValidEnd = isDate(line.substring(line.length() - DATE_LENGTH));              //valid end of line (dd/mm/yyyy)
+                if (currentLineHasValidStart && currentLineHasValidEnd) {               //line is complete
```

Comment:
I really don't understand the differences between the different bw.write(line). Can you explain it to me?

Reply:
I commented the same lines in different ways, forgot to check the comments.
br.write(line + "\r\n");  writes the line end goes to the next while without "\r\n" it stays on the same line.

Reply:
And this what kind of consequences have on the database?

Reply:
it makes it so each line of the text file is an entry:
Ref No,INCI name,INN name,Ph. Eur. Name,CAS No,EC No,
Chem/IUPAC Name / Description,Restriction,Function,Update Date

Reply:
If a line ends with a date it should be complete so it gets written and the line ends, otherwise the next line of the source file must be the part that is missing from the current line so, without ending the line, I read the next and write it continuing on the same line.